File: 1. Caesar Cipher/CaesarCipher.go
package main

import (
	"fmt"
	"math"
	"strings"
)

const TEXT = " ABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*(){<>?}0987654321"
const PRIVATE_KEY = 3

func main() {
	fmt.Printf("length is  %v \n", len(TEXT))
	inp := "cryptography and hashing!321"
	encrypt := encryption(inp)
	fmt.Println(encrypt)
	decrypt := decryption(encrypt)
	fmt.Println(strings.ToLower(decrypt))
}

func encryption(input string) string {
	msg := strings.ToUpper(input)
	cipher_text := ""
	// x = input
	// n = PRIVATE_KEY
	// En(x) = (x+n) % len(Text)
	for _, i := range msg {
		idx := strings.Index(TEXT, string(i))
		// encryption formula is
		f := (idx + PRIVATE_KEY) % len(TEXT)
		cipher_text += string(TEXT[f])
	}
	return cipher_text
}

func decryption(encypt string) string {
	plain_text := ""
	// x = encypt
	// n = PRIVATE_KEY
	// Dn(x) = (x-n) % len(TEXT)
	for _, i := range encypt {
		idx := strings.Index(TEXT, string(i))
		// decryption formula is
		f := (idx - PRIVATE_KEY) % len(TEXT)
		if f < 0 {
			// if f is negative value
			// f = f + len(TEXT)
			f = len(TEXT) - int(math.Abs(float64(f)))
		}
		plain_text += string(TEXT[f])
	}
	return plain_text
}

File: 1. Caesar Cipher/CaesarCipher.py
ALPHABET=" ABCDEFGHIJKLMNOPQRSTUVWXYZ!@#$%^&*(){<>?}"
PRIVATE_KEY = 3

# En(x) = (x+n) % len(ALPHABET) 
# x = character of plain text
# n = PRIVATE_KEY 

def caesar_cipher(plain_text):
    # encryption message
    cipher_text = ''
    plain_text = plain_text.upper()
    
    for chr in plain_text:
        # find the index of every character
        idx = ALPHABET.find(chr)
        # now the formula of caesar encryption is 
        # to the key use modular arithmetic to transform the values within 
        idx = (idx + PRIVATE_KEY) % len(ALPHABET)
        
        cipher_text += ALPHABET[idx]
    return cipher_text
        

# Caesar decryption
# formula  Dn(x) = (x - n) % len(ALPHABET)

def caesar_decypt(cipher_text):
    plain_text = ''
    for c in cipher_text:
        idx = ALPHABET.find(c)
        # Dn(x) = (x - n) % len(ALPHABET)
        idx = (idx - PRIVATE_KEY) % len(ALPHABET)
        plain_text += ALPHABET[idx]
    return plain_text

if __name__ == "__main__":
    # text = "cryptography and hashing?}"
    text = "hello"
    encryption = caesar_cipher(text)
    print(encryption)
    print(caesar_decypt(encryption).lower())